wordpress-mobile/WordPress-Android
# Update Apache Commons Text to 1.10.0

After an internal discussion (see `peaMlT-2f-p2` and `peaMlT-2f-p2#comment-78`), this PR upgrades Apache's `Commons Text` library to [1.10.0](https://commons.apache.org/proper/commons-text/changes-report.html#a1.10.0) in order fix the possible `StringSubstitutor` related vulnerability in the library.

For more details see: https://commons.apache.org/proper/commons-text/changes-report.html#a1.10.0

And more specifically see release notes entry below:

```
Make default string lookups configurable via system property. Remove
dns, url, and script lookups from defaults. If these lookups are
required for use in StringSubstitutor.createInterpolator(), they must be
enabled via system property. See StringLookupFactory for details.
```

-----

As part of this `commonsText = '1.10.0'` update the below changes were also applied in order to try and strip down as of the Apache Commons dependency as possible:

Remove `Apache Commons` from `Editor` module:

1. [Rename .java to .kt.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/9961c855909b571eb9feef712f64cb86d2cd82d6)
2. [Kotlinize metadata utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/eca153f80d5cf4870fab0dac51de6c2d7127451a)
3. [Update metadata utils kotlin to match its previous java style.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/510d2b0e13992fbd44102bc10d18371fa4954e34)
4. [Resolve detekt warnings for metadata utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/61afa49ac02db5025e035c0fbe32f2da1c230f2b)
5. [Replace unnecessary org apache commons lang dependency for editor.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/fe8246d2a7551e1dacd537ef27837ff020b44c93)

Strip down `Apache Commons` from `WordPress` module:

1. [Remove unnecessary commons lang not implemented exception.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/e950a413313086a4b0b104ec5dfc48d1b62b3352)
2. [Remove unnecessary commons text word utils capitalize.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/ed0c8f427bf5940b7dc38f121740de63cff89388)
3. [Replace commons lang array utils remove with array utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/48c87b341dd2dd3a3a43ac36226d27b0e38e9784)
4. [Remove unnecessary commons lang array utils remove.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/ce6fe7e499573d6e42f09d5166792999792e0c60)
5. [Replace commons lang array utils index of with array utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/762fb23046fdde78b92bd90687a56bf5736d915b)
6. [Remove unnecessary commons lang array utils index of.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/a2fb8535ebfa0e9cebc85a80208b8d749d10dbd2)
7. [Remove unnecessary commons lang date utils add and is same day.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/fcd2cae96c8b55aecde25d2359ac5bd1642b3fcb)

Suppression List:

1. [Suppress detekt warnings for metadata utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/9f4ab790ecbcc741c2b2b66cbf7c6506075ab7ea)
2. [Suppress magic number for stats date utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/066871776af867ed911b9d16fa22d16a6f7e8965)

Refactoring List:

1. [Rename stats related date utils to stats date utils.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/f1b475ef6d09fdb9ecac17b85ed91ea5948ba85c)
3. [Fix string escape utils import.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/4655c1ffb4854d79563dec1879e287f0b93cedda)
4. [Remove unnecessary remote url local variable.](https://github.com/wordpress-mobile/WordPress-Android/pull/17501/commits/b146be1fb904834b7c326e5fd3b03523c4e64caf)

-----

With this change the only Apache's `Commons Lang` and `Commons Text` related utilities that this project is using are the below:

1. `Commons Lang` -> `org.apache.commons.lang3.StringUtils`
    1. `StringUtils.isEmpty(...)` -> See `PostUtils.java` class.
    2. `StringUtils.equals(...)` -> See `PostUtils.java` class.
    3. `StringUtils.join(...)` -> See `MediaBlockType.java` class.
2. `Commons Text` -> `org.apache.commons.text.StringEscapeUtils`
    1. `StringEscapeUtils.unescapeHtml4(...)` -> 22 usages in project.

A quick diff of the previous usage of `commons-lang3-3.8.1` vs. the current usage of `commons-lang3-3.12.0` showed that only the `StringUtils.equals(...)` got updated, see `Step-wise comparison` below:

`commons-lang3-3.8.1`
```
    public static boolean equals(final CharSequence cs1, final CharSequence cs2) {
        ...
        return CharSequenceUtils.regionMatches(cs1, false, 0, cs2, 0, cs1.length());
    }
```

`commons-lang3-3.12.0`
```
    public static boolean equals(final CharSequence cs1, final CharSequence cs2) {
        ...
        // Step-wise comparison
        final int length = cs1.length();
        for (int i = 0; i < length; i++) {
            if (cs1.charAt(i) != cs2.charAt(i)) {
                return false;
            }
        }
        return true;
    }
```

A quick diff of the previous usage of `commons-text-1.1` vs. the current usage of `commons-text-1.10.0` showed that only the inner to `translate(...)` method of `StringEscapeUtils.unescapeHtml4(...)` got slightly updated, see `RuntimeException` vs. `UncheckedIOException` below:

`commons-text-1.1`
```
    public final String translate(final CharSequence input) {
        ...
        } catch (final IOException ioe) {
            // this should never ever happen while writing to a StringWriter
            throw new RuntimeException(ioe);
        }
    }
```

`commons-text-1.10.0`
```
    public final String translate(final CharSequence input) {
        ...
        } catch (final IOException ioe) {
            // this should never ever happen while writing to a StringWriter
            throw new UncheckedIOException(ioe);
        }
    }
```

-----

To test:

1. First, quickly smoke test both, the `WordPress` and `Jetpack` apps, and see if they work as expected.
2. Then, try and be more specific by taking a look at the code changes, associate those to some screens and make sure everything is working as expected, especially on the `Editor` side of things due to the `MetadataUtils` kotlinization.

-----

## Regression Notes

1. Potential unintended areas of impact

The `PostUtils.postHasEdits(...)` method might be producing different results before and after the change, that is, since the `StringUtils.equals(...)` implementation was slightly changed.

9. What I did to test those areas of impact (or what existing automated tests I relied on)

See `To test` section above.

10. What automated tests I added (or what prevented me from doing so)

N/A

-----

PR submission checklist:
